### PR TITLE
Prevent concurrent deletes for the same persistence id

### PR DESCRIPTION
**Pull request by chbatey**
Prevent concurrent deletes for the same persistence id

### CODE REVIEW

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala [hunk 1]
```diff
@@ -338,12 +373,179 @@ class CassandraJournal(cfg: Config) extends AsyncWriteJournal
...
   }
 
+  /**
+   * Not thread safe. Assumed to only be called from the journal actor.
+   * However, unlike asyncWriteMessages it can be called before the previous Future completes
+   */
+  override def asyncDeleteMessagesTo(persistenceId: String, toSequenceNr: Long): Future[Unit] = {
+
+    // TODO could "optimize" away deletes that overlap?
+    pendingDeletes.get(persistenceId) match {
```
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: This is called from the outside, not only from receive, so `pendingDeletes` will have to be thread safe, no? Probably easiest to defer it to work in `self` somehow.

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala [hunk 2]
```diff
@@ -338,12 +373,179 @@ class CassandraJournal(cfg: Config) extends AsyncWriteJournal
...
+      case null =>
+        log.debug("No outstanding delete for persistence id {}. Sequence nr: {}", persistenceId, toSequenceNr)
+        // fast path, no outstanding deletes for this persistenceId
+        val p = Promise[Unit]()
+        pendingDeletes.put(persistenceId, List(PendingDelete(persistenceId, toSequenceNr, p)))
```
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: these could be mixed in any order on two concurrent calls for the same `persistenceId` leading to one getting lost for example

File: core/src/test/scala/akka/persistence/cassandra/journal/CassandraJournalDeletionSpec.scala [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      probe.expectNoMessage()
+
+      p1 ! PoisonPill
+
+      // Recovery should not find a missing sequence nr
+      val p1TakeTwo = system.actorOf(Props(new PAThatDeletes("p1", probe.ref)))
+      p1TakeTwo ! PersistMe(101)
+      expectMsg(Ack)
+    }
+  }
```
Comment (johanandren) [inline core/src/test/scala/akka/persistence/cassandra/journal/CassandraJournalDeletionSpec.scala]: Cover that it fails if you overwhelm it?

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala (hunk 1 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: Any of the methods that are implemented from the `AsyncWriteJournal` are called on the actor thread of the journal actor, we rely on this inside `asyncWriteMessages` where we use a regular Map to track concurrent writes.

I moved the delete code from a separate trait to stop it being used outside of a `CassandraJournal`, we had moved it out as we thought we were going to use it in `akka-update` but we took it out of scope in the end.

File: core/src/test/scala/akka/persistence/cassandra/journal/CassandraJournalDeletionSpec.scala (hunk 3 above)
Comment (chbatey) [reply core/src/test/scala/akka/persistence/cassandra/journal/CassandraJournalDeletionSpec.scala]: Yep, will add a test for that now

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala (hunk 2 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: Don't think they can as per previous comment

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala (hunk 1 above)
Comment (johanandren) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: Ah ok, I was under the misconception that you actually get the journal instance back from the persistence API, good then.

Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: That's the case for the read side. Write journals are purely internal to akka-persitsence

File: core/src/main/resources/reference.conf [hunk 4]
```diff
@@ -92,6 +92,11 @@ cassandra-journal {
   # https://docs.datastax.com/en/developer/java-driver/3.1/manual/logging/#logging-query-latencies 
   log-queries = off
 
+  # Only one outstanding delete per persistenceId is permitted. Additional deletes are queued.
```
[the remark this answers is not in the document]
Comment (chbatey) [reply core/src/main/resources/reference.conf]: I'll re-write. Deletes are only enqueued if they are for the same persistence id.

[the remark this answers is not in the document]
Comment (chbatey) [reply core/src/main/resources/reference.conf]: # Akka-persistence allows multiple pending deletes for the same persistence id however this plugin only executes one
  # at a time per persistence id (deletes for different persistenceIds can happen concurrently).
  #
  # If multiple deletes for the same persistence id are received then they are queued.
  #
  # If the queue is full any subsequent deletes are failed immediately without attempting them in Cassandra.
  #
  # Deleting should be used with snapshots for efficiency for recovery thus deleting should be infrequent

**Pull request merged**